spec/index.bs

```diff
@@ -9918,34 +9929,49 @@ It must only be included by interfaces which also include those mixins.
                 [=Device timeline=] steps:
 
                 1. [$Validate the encoder state$] of |this|. If it returns false, stop.
-                1. If any of the following conditions are unsatisfied, make |this| [=invalid=] and stop.
+                1. Let |dynamicOffsetCount| be 0 if `bindGroup` is `null`, or
+                    |bindGroup|.{{GPUBindGroup/[[layout]]}}.{{GPUBindGroupLayout/[[dynamicOffsetCount]]}} if not.
+                1. If any of the following requirements are unmet, make |this| [=invalid=] and stop.
 
                     <div class=validusage>
-                        - |bindGroup| is [$valid to use with$] |this|.
-                        - |index| &lt; |this|.{{GPUObjectBase/[[device]]}}.{{device/[[limits]]}}.{{supported limits/maxBindGroups}}.
-                        - |dynamicOffsets|.length is
-                            |bindGroup|.{{GPUBindGroup/[[layout]]}}.{{GPUBindGroupLayout/[[dynamicOffsetCount]]}}.
+                        - |index| must be &lt;
+                            |this|.{{GPUObjectBase/[[device]]}}.{{device/[[limits]]}}.{{supported limits/maxBindGroups}}.
+                        - |dynamicOffsets|.length must be |dynamicOffsetCount|.
+                    </div>
+                1. If |bindGroup| is `null`:
+                    1. [=map/Remove=] |this|.{{GPUBindingCommandsMixin/[[bind_groups]]}}[|index|].
+                    1. [=map/Remove=] |this|.{{GPUBindingCommandsMixin/[[dynamic_offsets]]}}[|index|].
 
-                        - [$Iterate over each dynamic binding offset$] in |bindGroup| and
-                            run the following steps for each |bufferBinding|, |bufferLayout|,
-                            and |dynamicOffsetIndex|:
+                    Otherwise:
 
-                            - Let |bufferDynamicOffset| be |dynamicOffsets|[|dynamicOffsetIndex|].
-                            - |bufferBinding|.{{GPUBufferBinding/offset}} + |bufferDynamicOffset| +
-                                |bufferLayout|.{{GPUBufferBindingLayout/minBindingSize}} &le;
-                                |bufferBinding|.{{GPUBufferBinding/buffer}}.{{GPUBuffer/size}}.
-                            - if |bufferLayout|.{{GPUBufferBindingLayout/type}} is {{GPUBufferBindingType/"uniform"}}:
+                    1. If any of the following requirements are unmet, make |this| [=invalid=] and stop.
 
-                                - |dynamicOffset| is a multiple of {{supported limits/minUniformBufferOffsetAlignment}}.
+                        <div class=validusage>
+                            - |bindGroup| must be [$valid to use with$] |this|.
+                            - If |this| is a {{GPURenderCommandsMixin}}:
+                                - |bindGroupSpaceUsed| + |vertexBufferSpaceUsed| must be &le;
+                                    |this|.{{GPUObjectBase/[[device]]}}.{{device/[[limits]]}}.{{supported limits/maxBindGroupsPlusVertexBuffers}}, where:
```

Comment:
This structure (validate X where X is ...) was a bit surprising. We don't use that anywhere else in the spec that I'm aware of? Not a blocking issue for me, it just read a little weird.

Also, I was surprised that we're doing the validation for this at binding time, since I was under the impression that it was draw time. (I probably just missed that part of the editors meeting.) Keeping things out of the draw validation seems good, it just enforces a stricter sequence of events than I anticipated. Again, not a blocker.

Reply:
I believe @litherum's design motivation required that this be validated at binding time. Validating at draw time means buffering up Metal binding commands until draw time, because they might be invalid. And if you're going to do that anyway, then you could just not add this validation at all and rely only on the validation in pipeline creation.

Reply:
> This structure (validate X where X is ...) was a bit surprising. We don't use that anywhere else in the spec that I'm aware of? Not a blocking issue for me, it just read a little weird.

In other places we have "Let" directives inside of validation rule blocks. This doesn't make sense because those blocks are unordered, so you can't define a variable in one and use it in the next. However I could just write some ordered steps inside here so I think I'll do that.

Reply:
Can't figure out a way that doesn't look funny so actually will leave it for now.

![image](https://user-images.githubusercontent.com/606355/226068506-44eb12d5-fa40-4cb4-b2d4-1a1e52455393.png)

Reply:
Editors: Move validation to draw time, as originally planned. Putting it here is more consistent with all of the validations that already exist between pipeline and everything else. (AI for me to update this.)